aspnetcore/fundamentals/http-requests/samples/GitHub/GitHubService.cs

```diff
@@ -0,0 +1,52 @@
+using System;
+using System.IO;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+
+namespace HttpClientFactorySample.GitHub
+{
+    /// <summary>
+    /// Exposes methods to return GitHub API data
+    /// </summary>
+    #region snippet1
+    public class GitHubService
+    {
+        public HttpClient Client { get; }
+
+        public GitHubService(HttpClient client)
+        {
+            client.BaseAddress = new Uri("https://api.github.com/");
+            client.DefaultRequestHeaders.Add("Accept", "application/vnd.github.v3+json"); // GitHub API versioning
+            client.DefaultRequestHeaders.Add("User-Agent", "HttpClientFactory-Sample"); // GitHub requires a user-agent
+
+            Client = client;
+        }
+
+        public async Task<GitHubIssue> GetLatestDocsIssue()
+        {
+            using (var stream = await Client.GetStreamAsync("/repos/aspnet/docs/issues?state=open&sort=created&direction=desc"))
+            using (var streamReader = new StreamReader(stream))
+            using (var jsonReader = new JsonTextReader(streamReader))
+            {
+                var serializer = new JsonSerializer();
+
+                while (await jsonReader.ReadAsync())
+                {
+                    if (jsonReader.TokenType == JsonToken.StartObject)
+                    {
+                        var issue = serializer.Deserialize<GitHubIssue>(jsonReader);
+
+                        if (issue != null)
+                        { 
+                            return issue; // we only want the first object
+                        }
+                    }
```

Comment:
This code should be using `ReadAsAsync<>` https://github.com/rynowak/Samples/blob/dev/MusicStore/src/MusicStoreUI/Services/MusicStoreService.cs#L37

Reply:
Handy! I wasn't aware ReadAsAsync was available. I believe (while more complex) the code I have is slightly more efficient? In this example it's taking only the first item. A bit of a contrived example. With ReadAsAsync I get all items back in the list. A very rough comparison is ~150kb extra allocated between memory snapshots.

I spoke to @glennc and @DamianEdwards at Summit as I was concerned about this complexity and if in fact this was a reasonable approach. At that time it seemed like we wanted to include something similar.

I'm happy to make the change though?

Side query: Am I correct in thinking that https://github.com/aspnet/AspNetWebStack/blob/master/src/System.Net.Http.Formatting/HttpContentExtensions.cs is the source building Microsoft.AspNet.WebApi.Client?

Reply:
If what you're saying is true then we should choose a different example. The focus of this doc isn't on the minutae of how to implement special case handling of JSON scenarios, so it should use the idiomatic approach - which is `ReadAsAsync`.

Reply:
I can adjust the sample to return the full list and switch to ReadAsAsync. Not a problem from my side so I'll get that done hopefully later today / tomorrow.